src/optimizer/expression_heuristics.cpp

```diff
@@ -0,0 +1,233 @@
+#include "duckdb/optimizer/expression_heuristics.hpp"
+#include "duckdb/planner/expression/list.hpp"
+
+using namespace duckdb;
+using namespace std;
+
+unique_ptr<LogicalOperator> ExpressionHeuristics::Rewrite(unique_ptr<LogicalOperator> op) {
+    VisitOperator(*op);
+    return op;
+}
+
+void ExpressionHeuristics::VisitOperator(LogicalOperator &op) {
+    if (op.type == LogicalOperatorType::FILTER) {
+        //reorder all filter expressions
+		if (op.expressions.size() > 1) {
+			ReorderExpressions(op.expressions);
+		}
+    }
+
+    //traverse recursively through the operator tree
+    VisitOperatorChildren(op);
+}
+
+void ExpressionHeuristics::ReorderExpressions(vector<unique_ptr<Expression>> &expressions) {
+
+	struct expr {
+		index_t idx;
+		index_t cost;
+
+		bool operator==(const expr &p) const {
+		return idx == p.idx && cost == p.cost;
+		}
+		bool operator<(const expr &p) const {
+		return cost < p.cost || (cost == p.cost && idx < p.idx);
+		}
+	};
+
+	vector<expr> exprCosts;
+	//iterate expressions, get cost for each one and order by that cost
+	for (index_t i = 0; i < expressions.size(); i++) {
+		exprCosts.push_back({i, Cost(*expressions[i])});
+	}
+
+	sort(exprCosts.begin(), exprCosts.end());
+	assert(exprCosts.size() == expressions.size());
+
+	//for all elements to put in place
+	for(index_t i = 0; i < expressions.size() - 1; ++i) {
+		//while the element i is not yet in place
+		while(i != exprCosts[i].idx) {
+			//swap it with the element at its final place
+			index_t alt = exprCosts[i].idx;
+			swap(expressions[i], expressions[alt]);
+			swap(exprCosts[i], exprCosts[alt]);
+		}
+	}
+}
+
+index_t ExpressionHeuristics::Cost(Expression &expr) {
+	switch (expr.expression_class) {
+		case ExpressionClass::BOUND_CASE: {
+			auto &case_expr = (BoundCaseExpression &)expr;
+
+			//CASE WHEN check THEN result_if_true ELSE result_if_false END
+			return Cost(*case_expr.check) + Cost(*case_expr.result_if_true) + Cost(*case_expr.result_if_false) + 5;
+		}
+		case ExpressionClass::BOUND_CAST: {
```

Comment:
I would suggest adding functions for each of the expression types instead of making a very large switch statement. Then in the switch statement, you only call the function, e.g. for the cast expression, you would add the following function:

```cpp
index_t ExpressionCost(BoundCastExpression &expr) {
	//determine cast cost by comparing cast_expr.source_type and cast_expr_target_type
	if (cast_expr.target_type != cast_expr.source_type) {
		//if cast from or to varchar
		//TODO: we might want to add more cases
		if (cast_expr.target_type == SQLType::VARCHAR || cast_expr.source_type == SQLType::VARCHAR) {
			return 200;
		} else {
			return 5;
		}
	}
    return Cost(*cast_expr.child) + cast_cost;
}
```

Then in the switch statement, you only have the following smaller piece of code:

```cpp
auto &cast_expr = (BoundCastExpression &)expr;
return ExpressionCost(cast_expr);
```